platform/projectModel-api/src/com/intellij/openapi/roots/DependencyScope.java

```diff
@@ -97,6 +97,23 @@ public boolean isForTestRuntime() {
     return myForTestRuntime;
   }
 
+  public @NotNull DependencyScope including(@NotNull DependencyScope other) {
+    if (this == other || this.includes(other)) return this;
+    if (other.includes(this)) return other;
+    // At this point, we know that `this` and `other` are different, neither is COMPILE, and they're not PROVIDED+TEST either.
+    // The remaining cases are:
+    // - PROVIDED+RUNTIME = COMPILE
+    // - TEST+RUNTIME = COMPILE (broader than necessary, but there's no closer match)
+    return COMPILE;
+  }
+
+  public boolean includes(@NotNull DependencyScope other) {
+    return (myForProductionCompile || !other.myForProductionCompile) &&
+           (myForProductionRuntime || !other.myForProductionRuntime) &&
+           (myForTestCompile || !other.myForTestCompile) &&
+           (myForTestRuntime || !other.myForTestRuntime);
+  }
+
```

Comment:
I can of course move this functionality to an external utility class, but the main reason why I thought it belonged here is for better maintainability through encapsulation. If this is implemented externally, and the `DependencyScope` enum ever changes (e.g. modification of the meaning of an existing scope, or the addition of an extra property or even a new scope), the developer making that change will have a much lower chance of noticing that `GradleProjectResolverUtil` also needs to be changed. Is there perhaps a better name for these methods, that would be suitable to have here in the enum itself?

If you still prefer I move this out of the enum, let me know and I'll do that.

Reply:
Encapsulation is a good reason. But indeed, we need a better name then.
I need to better understand the purpose of the method. At the first glance, it tries to _enlarge_ the current scope so that current dependencies are included whenever the _other scope_ or _current scope_ dependencies would be included (according to the scope rules). Is that true? Can it be a hint for a new name?

Reply:
* `scope1.includes(scope2)` is `true` if and only if `scope1` "applies" in all of the cases when `scope2` does. So if `scope2.isForProductionCompile()` is `true`, then `scope1.isForProductionCompile() ` must also be `true` in order for `scope1.includes(scope2)` to be `true`.
* `scope1.including(scope2)` is meant to return the "narrowest" possible scope that `includes` both `scope1` and `scope2`. So basically yes, it's about "widening" the scope, but only as much as necessary. Many times, it will actually return `scope1` or `scope2`, because one already "applies" in all of the cases when the other would.

The reason I'm having a hard time finding a suitable and descriptive name is because the terminology is not really clear to me with regards to scopes. There's only class level Javadocs, but even that starts by saying "_The table below specifies which order entries are used during compilation and runtime_", even though the class seems to have nothing to do with ordering, but rather acts as a filter on dependencies, controlling which ones get onto the classpath in certain use cases (such as "Production Compile").